### PR TITLE
Lua scripting engine

### PR DESCRIPTION
**Pull request by ja2-derek**
Basic implementation of #1205.

_Ignore all files under dependencies/lib-lua, and the diff should be more managable._

# Lua

Lua (v5.3.5) and Sol2 (v2.20.6) are bundled and built from source. System packages are not considered at all.  Lua is a small library, and the pre-compiled Lua binaries did not work for me (VC++), probably because of Debug/Release or VS version differences. But the compilation of Lua is quite fast, and always building from source simplifies the CMakeFile quite a bit

**sol2 requires C++ 14**

# Scripting Engine

There are two main module for this so-called "scripting engine". First is the functions available to Lua scripts (FunctionsLibraries.h), and second is the integration points where C code will call lua function if defined (see ScriptingExtensions.h).

Mods using Lua must provide a `data/scripts/main.lua`. This file is the entrypoint (think `int main`), and can include other lua files. The base game will also provide some constants, enums or maybe helper functions in Lua.

The functions exposed to Lua are a slightly simplified versions of the corresponding C functions, and should be restricted to use only by Lua scripts. The functions are all grouped in the `scripting/` module. I intend to generate some form of documentations later on.

# Error handling

All Lua scripts are fully parsed before game starts. If there is an error (e.g. syntax error), an exception is thrown and the game will refuse to start.

If Lua script had a runtime error (within Lua, or within a C function called by Lua), the Lua error will be logged and thrown as `std::runtime_error`. It may or may not be handled by the game depending where it is thrown. The scripting engine will be disabled until the game restarts.

# Example implementation

The `generous-rebel` minimod is updated to use Lua to extend its functionality. It now adds a few items to the initial sectors at the first visit.

### CODE REVIEW

**Comment by ja2-derek** [comment]
A (very basic) manual for using scripting in JA2:S mods: [scripting-manual.pdf](https://github.com/ja2-stracciatella/ja2-stracciatella/files/5177403/scripting-manual.pdf)

**Comment by lynxlynxlynx** [comment]
Did you make that manually or is there some doxygen, sphinx or similar behind it?

**Comment by ja2-derek** [comment]
That's generated with Doxygen from annotated comments in code

File: src/externalized/scripting/ScriptingExtensionsLua.cc [hunk 1]
```diff
@@ -128,15 +129,18 @@ static void RegisterGlobals()
...
 	lua.set_function("GetUndergroundSectorInfo", GetUndergroundSectorInfo);
 	
 	lua.set_function("CreateItem", CreateItem);
 	lua.set_function("CreateMoney", CreateMoney);
 	lua.set_function("PlaceItem", PlaceItem);
+
+	lua.set_function("JA2Require", JA2Require);
+	lua.set_function("require",  [](void) { throw std::logic_error("require is not allowed. Use JA2Require instead"); });
+	lua.set_function("dofile",   [](void) { throw std::logic_error("require is not allowed. Use JA2Require instead"); });
+	lua.set_function("loadfile", [](void) { throw std::logic_error("require is not allowed. Use JA2Require instead"); });
```
Comment (lynxlynxlynx) [inline src/externalized/scripting/ScriptingExtensionsLua.cc:143]: These two have the wrong name in the error.

**Pull request merged**